HIP/hip-376.md

```diff
@@ -0,0 +1,156 @@
+- `function approve(address spender, uint256 amount) external returns (bool);`
+- `function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);`
+
+#### Supported ERC-721 operations
+
+Tokens of type `NON_FUNGIBLE_UNIQUE` will support standard `ERC-721` calls.
+
+The following `ERC-721` operations will be supported. Standard `ERC-721 Events` will be emitted as appropriate.
+
+  - `function approve(address _approved, uint256 _tokenId) external payable`
+  - `function setApprovalForAll(address _operator, bool _approved) external`
+  - `function getApproved(uint256 _tokenId) external view returns (address)`
+  - `function isApprovedForAll(address _owner, address _operator) external view returns (bool)`
+  - `function transferFrom(address _from, address _to, uint256 _tokenId) external payable`
+
+#### ERC20 and ERC721 mapping to the HAPI interface
+
+There are several differences in the way `ERC20` and `ERC721` functions are designed to behave and the way HAPI transactions for `approve`, `allowance` and `transferFrom` are implemented in Hedera. This section describes the integration between the function specifications and the actual Hedera behaviour.
+
+##### ERC20 approve and HAPI Adjust allowance
+
+ERC20 `approve` function works by overriding the previous value (if any was set) for the 'spender' address specified.
+HAPI `CryptoAdjustAllowance` transactions work by applying the difference specified by the `amount` field in the transaction body to the existing allowance (or 0 if there is no existing allowance approved for that account).
+Calling ERC20 `approve` with amount `200` will be implemented in the following way in the precompile:
+ - query the current configured allowance for the `owner`, `spender` and `tokenId`
+ - calculate the adjustment that needs to be made to have the final allowance equal to the amount received as argument to the ERC20 `approve`
+ - create a `CryptoAdjustAllowance` tx with the calculated adjustment
+
+##### ERC721 approve and multiple approved accounts for NFT in Hedera
+
+ERC721 is designed in such a way that an NFT can have either 0 or 1 approvals e.g. if Alice owns NFT 'X' and Alice calls the approve function two times, once with Bob's account and once with Carol's account, the second call to approve will override Bob's allowance and only Carol will have approved allowance for that NFT. In Hedera the behaviour is different and both Bob and Carol would have approved allowances after the second call to approve.
+To make sure there is no unexpected result for smart contract developers working with HTS tokens, ERC721 `approve` and `getApproved` will behave in the following way with HTS tokens:
+When `approve` is called:
+ - first we'll check if the address from the function arguments is the zero address
+ - if yes, then we'll iterate through the account's `GrantedNFTAllowances` and revoke (adjust with -) any existing approvals for the NFT serial number
```

Comment:
From [HIP-336](https://hips.hedera.com/hip/hip-336): Each account is limited to 100 allowances. This limit spans hbar, fungible and non-fungible token allowances. Each NFT serial number will count as a single allowance when computing the total allowances an account currently has.

I agree with the concern you expressed. As I see it we have two options:

1.  Calculate gas price based on number of `GrantedNFTAllowances` we need to revoke
2. Only revoke allowance if there's exactly 1 granted, otherwise revert and require HAPI manipulation of approvals

Personally I like option 1 better since option 2 seems to be a bit limiting, if the user **knows** he'd like to revoke all grants.